<h1>Tabla de Contenidos<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Inicialización" data-toc-modified-id="Inicialización-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Inicialización</a></span></li><li><span><a href="#Conección-a-la-Base-de-Datos" data-toc-modified-id="Conección-a-la-Base-de-Datos-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Conección a la Base de Datos</a></span></li><li><span><a href="#Descripción-de-los-datos" data-toc-modified-id="Descripción-de-los-datos-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Descripción de los datos</a></span></li><li><span><a href="#Ejercicios" data-toc-modified-id="Ejercicios-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Ejercicios</a></span><ul class="toc-item"><li><span><a href="#Número-de-libros-publicados-después-del-1-de-enero-de-2000" data-toc-modified-id="Número-de-libros-publicados-después-del-1-de-enero-de-2000-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Número de libros publicados después del 1 de enero de 2000</a></span></li><li><span><a href="#Número-de-reseñas-de-usuarios-y-la-calificación-promedio-para-cada-libro" data-toc-modified-id="Número-de-reseñas-de-usuarios-y-la-calificación-promedio-para-cada-libro-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Número de reseñas de usuarios y la calificación promedio para cada libro</a></span></li><li><span><a href="#Identificar-la-editorial-que-ha-publicado-el-mayor-número-de-libros-con-más-de-50-páginas" data-toc-modified-id="Identificar-la-editorial-que-ha-publicado-el-mayor-número-de-libros-con-más-de-50-páginas-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>Identificar la editorial que ha publicado el mayor número de libros con más de 50 páginas</a></span></li><li><span><a href="#Identificar-al-autor-que-tiene-la-más-alta-calificación-promedio-del-libro:-mirar-solo-los-libros-con-al-menos-50-calificaciones" data-toc-modified-id="Identificar-al-autor-que-tiene-la-más-alta-calificación-promedio-del-libro:-mirar-solo-los-libros-con-al-menos-50-calificaciones-4.4"><span class="toc-item-num">4.4&nbsp;&nbsp;</span>Identificar al autor que tiene la más alta calificación promedio del libro: mirar solo los libros con al menos 50 calificaciones</a></span></li><li><span><a href="#Encontrar-el-número-promedio-de-reseñas-de-texto-entre-los-usuarios-que-calificaron-más-de-50-libros" data-toc-modified-id="Encontrar-el-número-promedio-de-reseñas-de-texto-entre-los-usuarios-que-calificaron-más-de-50-libros-4.5"><span class="toc-item-num">4.5&nbsp;&nbsp;</span>Encontrar el número promedio de reseñas de texto entre los usuarios que calificaron más de 50 libros</a></span></li></ul></li></ul></div>

<a id='contenidos'></a>

# Descripción del proyecto  
El coronavirus tomó al mundo entero por sorpresa, cambiando la rutina diaria de todos y todas. Los habitantes de las ciudades ya no pasaban su tiempo libre fuera, yendo a cafés y centros comerciales; sino que más gente se quedaba en casa, leyendo libros. Eso atrajo la atención de las startups (empresas emergentes) que se apresuraron a desarrollar nuevas aplicaciones para los amantes de los libros.

Tenemos una base de datos de uno de los servicios que compiten en este mercado. Contiene datos sobre libros, editoriales, autores y calificaciones de clientes y reseñas de libros. Esta información se utilizará para generar una propuesta de valor para un nuevo producto.

[Volver a contenidos](#contenidos)
## Inicialización 
En esta sección vamos a cargar las librerías para el estudio

In [1]:
# importar librerías
import pandas as pd
from sqlalchemy import create_engine, inspect
import os
from dotenv import load_dotenv
load_dotenv()

True

[Volver a contenidos](#contenidos)
## Conección a la Base de Datos 
En esta sección vamos a realizar una conección a la BD para obtener información de diversas tablas.

In [2]:
# coneccion a la BD

db_config = {'user':os.getenv('tripleten_sql_user'),         # nombre de usuario
             'pwd': os.getenv('tripleten_sql_pwd'), # contraseña
             'host': os.getenv('tripleten_sql_host'),
             'port': 6432,              # puerto de conexión
             'db': os.getenv('tripleten_sql_db')}          # nombre de la base de datos

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                                     db_config['pwd'],
                                                                       db_config['host'],
                                                                       db_config['port'],
                                                                       db_config['db'])

engine = create_engine(connection_string, connect_args={'sslmode':'require'},pool_size=10, max_overflow=20)

In [3]:
# revisar los esquemas en la BD
inspector=inspect(engine)
schemas=inspector.get_schema_names()

for schema in schemas:
    print("schema: %s" % schema)
    for table_name in inspector.get_table_names(schema=schema):
        print(f'table_name: ',table_name, f'----------------------------')
        for column in inspector.get_columns(table_name,schema=schema):
            print("Column: %s" % column)

schema: information_schema
table_name:  sql_features ----------------------------
Column: {'name': 'feature_id', 'type': VARCHAR(), 'nullable': True, 'default': None, 'autoincrement': False, 'comment': None}
Column: {'name': 'feature_name', 'type': VARCHAR(), 'nullable': True, 'default': None, 'autoincrement': False, 'comment': None}
Column: {'name': 'sub_feature_id', 'type': VARCHAR(), 'nullable': True, 'default': None, 'autoincrement': False, 'comment': None}
Column: {'name': 'sub_feature_name', 'type': VARCHAR(), 'nullable': True, 'default': None, 'autoincrement': False, 'comment': None}
Column: {'name': 'is_supported', 'type': VARCHAR(), 'nullable': True, 'default': None, 'autoincrement': False, 'comment': None}
Column: {'name': 'is_verified_by', 'type': VARCHAR(), 'nullable': True, 'default': None, 'autoincrement': False, 'comment': None}
Column: {'name': 'comments', 'type': VARCHAR(), 'nullable': True, 'default': None, 'autoincrement': False, 'comment': None}
table_name:  sql_imp

Las tablas de nuestro interés están en el **'schema'** llamado **'public'.**

In [4]:
# creamos una consulta a la BD para ver que podamos extraer información
query="SELECT * FROM public.books LIMIT 5"

In [5]:
# realizamos una consulta de prueba a la BD
pd.io.sql.read_sql(query, con = engine)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

[Volver a contenidos](#contenidos)
## Descripción de los datos

Las tablas de datos y las columnas que contienen cada una se describen a continuación.

In [6]:
# creamos una consulta a la tabla 'books'
query="SELECT * FROM public.books LIMIT 5"

# realizamos la consulta a la BD
pd.io.sql.read_sql(query, con = engine)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

**`books`**

Contiene datos sobre libros:

- `book_id`: identificación del libro
- `author_id`: identificación del autor o autora
- `title`: título
- `num_pages`: número de páginas
- `publication_date`: fecha de la publicación
- `publisher_id`: identificación de la editorial
---

In [7]:
# creamos una consulta a la tabla 'authors'
query="SELECT * FROM public.authors LIMIT 5"

# realizamos la consulta a la BD
pd.io.sql.read_sql(query, con = engine)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

**`authors`**

Contiene datos sobre autores:

- `author_id`: identificación del autor o autora
- `author`: el autor o la autora
---

In [8]:
# creamos una consulta a la tabla 'publishers'
query="SELECT * FROM public.publishers LIMIT 5"

# realizamos la consulta a la BD
pd.io.sql.read_sql(query, con = engine)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

**`publishers`**

Contiene datos sobre editoriales:

- `publisher_id`: identificación de la editorial
- `publisher`: la editorial
---

In [9]:
# creamos una consulta a la tabla 'ratings'
query="SELECT * FROM public.ratings LIMIT 5"

# realizamos la consulta a la BD
pd.io.sql.read_sql(query, con = engine)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

**`ratings`**

Contiene datos sobre las calificaciones de usuarios:

- `rating_id`: identificación de la calificación
- `book_id`: identificación del libro
- `username`: el nombre del usuario que revisó el libro
- `rating`: calificación
---

In [10]:
# creamos una consulta a la tabla 'reviews'
query="SELECT * FROM public.reviews LIMIT 5"

# realizamos la consulta a la BD
pd.io.sql.read_sql(query, con = engine)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

**`reviews`**

Contiene datos sobre las reseñas de los y las clientes:

- `review_id`: identificación de la reseña
- `book_id`: identificación del libro
- `username`: el nombre del usuario que revisó el libro
- `text`: el texto de la reseña
---

**Las relaciones existentes entre las tablas las podemos ver en el siguiente diagrama.**

![Relaciones_tablas](relaciones_tablas.png)

[Volver a contenidos](#contenidos)
## Ejercicios

### Número de libros publicados después del 1 de enero de 2000

In [11]:
# creamos una consulta a la tabla 'books'
query="SELECT COUNT(book_id) FROM public.books WHERE publication_date>'2000-01-01'"

# realizamos la consulta a la BD
count_books=pd.io.sql.read_sql(query, con = engine)
print(f'Despues del 01 de enero de 2000 se publicaron: ', count_books.iloc[0,0],' libros')

Despues del 01 de enero de 2000 se publicaron:  819  libros


### Número de reseñas de usuarios y la calificación promedio para cada libro

In [12]:
# creamos una consulta a la tabla 'books' y sus uniones con 'ratings' y 'reviews'
query="""
SELECT b.book_id, COUNT(r.review_id) AS conteo_review_id, AVG(ra.rating) AS promedio_rating
FROM public.books b
JOIN public.reviews r ON b.book_id = r.book_id
JOIN public.ratings ra ON b.book_id = ra.book_id
GROUP BY b.book_id;
"""
review_rating_by_book=pd.io.sql.read_sql(query, con = engine)
review_rating_by_book.head()

book_id  conteo_review_id  promedio_rating
0      652                 4         4.500000
1      273                 4         4.500000
2       51                60         4.250000
3      951                 4         4.000000
4      839                28         4.285714

El dataframe `review_rating_by_book` contiene el número de reseñas de usuarios y la calificación promedio para cada libro. 

### Identificar la editorial que ha publicado el mayor número de libros con más de 50 páginas

In [13]:
# creamos una consulta a la tabla 'books' y su union con 'publisher'
query="""
SELECT publisher, COUNT(book_id) AS conteo
FROM public.books b
JOIN public.publishers p ON b.publisher_id = p.publisher_id
WHERE num_pages > 50
GROUP BY publisher
ORDER BY conteo DESC
LIMIT 10;
"""
publisher=pd.io.sql.read_sql(query, con = engine)
publisher

publisher  conteo
0              Penguin Books      42
1                    Vintage      31
2   Grand Central Publishing      25
3           Penguin Classics      24
4           Ballantine Books      19
5                     Bantam      19
6                    Berkley      17
7         St. Martin's Press      14
8              Berkley Books      14
9  William Morrow Paperbacks      13

La editorial `Penguin Books` es la editorial que ha publicado el mayor número de libros con más de 50 páginas.

### Identificar al autor que tiene la más alta calificación promedio del libro: mirar solo los libros con al menos 50 calificaciones

In [14]:
# creamos una consulta a la tabla 'books' y sus uniones con 'authors' y 'ratings'
query="""
SELECT a.author, AVG(ra.rating) AS promedio_rating, COUNT(ra.book_id) AS conteo
FROM public.books b
JOIN public.authors a ON b.author_id = a.author_id
JOIN public.ratings ra ON b.book_id = ra.book_id
WHERE b.book_id IN (
                    SELECT book_id
                    FROM public.ratings
                    GROUP BY book_id
                    HAVING COUNT(book_id) >= 50
                    )
GROUP BY a.author
ORDER BY promedio_rating DESC
LIMIT 10;
"""
author=pd.io.sql.read_sql(query, con = engine)
author

author  promedio_rating  conteo
0                         J.K. Rowling/Mary GrandPré         4.287097     310
1                  Markus Zusak/Cao Xuân Việt Khương         4.264151      53
2                                     J.R.R. Tolkien         4.246914     162
3                                  Louisa May Alcott         4.192308      52
4                                       Rick Riordan         4.080645      62
5                                    William Golding         3.901408      71
6                                      J.D. Salinger         3.825581      86
7           Paulo Coelho/Alan R. Clarke/Özdemir İnce         3.789474      57
8  William Shakespeare/Paul Werstine/Barbara A. M...         3.787879      66
9                                         Lois Lowry         3.750000      56

Los autores `J.K. Rowling/Mary GrandPré ` son los que tienen la más alta calificación promedio del libro, para libros con al menos 50 calificaciones.

### Encontrar el número promedio de reseñas de texto entre los usuarios que calificaron más de 50 libros

In [15]:
# creamos una consulta a la tabla 'reviews' y una subconsulta a la tabla 'ratings'
query="""
SELECT AVG(tabla1.conteo_text) AS promedio_conteo_text
FROM (
        SELECT COUNT(r.text) AS conteo_text
        FROM public.reviews r
        WHERE r.username IN (
                            SELECT username
                            FROM ratings
                            GROUP BY username
                            HAVING COUNT(DISTINCT book_id) > 50
                            )
       GROUP BY username
    ) AS tabla1
"""
promedio_reviews=pd.io.sql.read_sql(query, con = engine)
promedio_reviews

promedio_conteo_text
0             24.333333

24.33 es el número promedio de reseñas de texto entre los usuarios que calificaron más de 50 libros.